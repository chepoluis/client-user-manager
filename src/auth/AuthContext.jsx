import { createContext, useContext, useState, useEffect } from "react";
import { useCheckAuth } from "../hooks/useCheckout";

const AuthContext = createContext();

export const useAuth = () => {
  return useContext(AuthContext);
};

export const AuthProvider = ({ children }) => {
  const status = useCheckAuth();
  const [isAuthenticated, setIsAuthenticated] = useState(false);

  useEffect(() => {
    if (status === "authenticated") {
      setIsAuthenticated(true);
    } else {
      setIsAuthenticated(false);
    }
  }, [status]);

  return (
    <AuthContext.Provider value={{ isAuthenticated, status }}>
      {children}
    </AuthContext.Provider>
  );
};
