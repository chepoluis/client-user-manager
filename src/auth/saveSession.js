// Function to save user information in session storage
export function saveUserToSessionStorage(user) {
  let userJson = JSON.stringify(user);

  sessionStorage.setItem("user", userJson);
  //   localStorage.setItem("user", userJson);
}

// Function to get user information from session storage
export function getUserFromSessionStorage() {
  let userJson = sessionStorage.getItem("user");
  //   let userJson = localStorage.getItem("user");

  let user = JSON.parse(userJson);

  return user;
}

export function deleteUserFromSessionStorage() {
  sessionStorage.removeItem("user");
}
