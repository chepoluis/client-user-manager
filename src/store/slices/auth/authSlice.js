import { createSlice } from "@reduxjs/toolkit";

export const authSlice = createSlice({
  name: "auth",
  initialState: {
    status: "not-authenticated", // 'checking' 'not-authenticated', 'authenticated'
    email: null,
    firstName: null,
    role: null,
    errorMessage: null,
  },
  reducers: {
    login: (state, { payload }) => {
      state.status = "authenticated";
      state.email = payload.email;
      state.firstName = payload.firstName;
      state.role = payload.role;
      state.errorMessage = null;
    },
    logout: (state, { payload }) => {
      state.status = "not-authenticated";
      state.email = null;
      state.firstName = null;
      state.role = null;
      state.errorMessage = payload?.errorMessage;
    },
    checkingCredentials: (state) => {
      state.status = "checking";
      state.errorMessage = null;
    },
  },
});

export const { login, logout, checkingCredentials } = authSlice.actions;
