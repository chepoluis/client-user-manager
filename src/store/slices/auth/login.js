import { mockDataUser } from "../../../data/mockData";

export const signInWithEmail = ({ email, password }) => {
  return new Promise((resolve, reject) => {
    setTimeout(() => {
      const user = mockDataUser.find(
        (user) => user.email === email && user.password === password
      );
      
      if (user) {
        resolve(user);
      } else {
        reject(false);
      }
    }, 2000);
  });
};
