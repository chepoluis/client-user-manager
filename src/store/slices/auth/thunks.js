import { saveUserToSessionStorage } from "../../../auth/saveSession";
import { checkingCredentials, login, logout } from "./authSlice";
import { signInWithEmail } from "./login";

export const startSignIn = ({ email, password }) => {
  return async (dispatch) => {
    try {
      dispatch(checkingCredentials());

      const user = await signInWithEmail({ email, password });

      saveUserToSessionStorage(user);

      const { email: userEmail, displayName } = user;
      
      dispatch(login({ email: userEmail, displayName }));
    } catch (err) {
      return dispatch(logout({ errorMessage: "Invalid email or password" }));
    }
  };
};

export const startLogout = () => {
  return async (dispatch) => {
      dispatch( logout({}) );
  }
}
