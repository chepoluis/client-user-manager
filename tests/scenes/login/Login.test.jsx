import "@testing-library/jest-dom/extend-expect";
import { configureStore } from "@reduxjs/toolkit";
import { fireEvent, render, screen, act } from "@testing-library/react";
import { Provider } from "react-redux";
import { MemoryRouter } from "react-router-dom";
import Login from "../../../src/scenes/login";
import { authSlice } from "../../../src/store/slices/auth/authSlice";
import { ThemeProvider, createTheme } from '@mui/material/styles';

const theme = createTheme();

const store = configureStore({
  reducer: {
    auth: authSlice.reducer,
  },
});

describe("tests in <Login />", () => {
  test("should match the snapshot", () => {
    const { asFragment } = render(
      <Provider store={store}>
        <ThemeProvider theme={theme}>
          <Login />
        </ThemeProvider>
      </Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test("should show the component correctly", () => {
    render(
      <Provider store={store}>
        <MemoryRouter>
          <Login />
        </MemoryRouter>
      </Provider>
    );

    expect(screen.getAllByText(/login/i).length).toBeGreaterThanOrEqual(1);
  });

  test("should display error message when submitting empty form", async () => {
    render(
      <Provider store={store}>
        <MemoryRouter>
          <Login />
        </MemoryRouter>
      </Provider>
    );

    const submitButton = screen.getByRole("button", { type: /submit/i });
    // eslint-disable-next-line testing-library/no-unnecessary-act
    await act(async () => {
      fireEvent.click(submitButton);
    });

    // const emailError = await screen.findByText(/Alert/i);
    const passwordHelperText = screen.getByTestId("password-helper-text");

    // expect(emailError).toBeInTheDocument();
    // expect(passwordError).toBeInTheDocument();
    // expect(passwordHelperText).toHaveTextContent("No password provided");
  });

  // test("should display error message when submitting empty form", () => {
  //   render(
  //     <Provider store={store}>
  //       <MemoryRouter>
  //         <Login />
  //       </MemoryRouter>
  //     </Provider>
  //   );

  //   const submitButton = screen.getByRole('button', { type: /submit/i });
  //   fireEvent.click(submitButton);

  //   expect(screen.getAllByText(/No email provided./i).length).toBeInTheDocument();
  //   expect(screen.getByText(/No password provided./i)).toBeInTheDocument();
  // });

  // test("should display error message when entering invalid email", () => {
  //   render(
  //     <Provider store={store}>
  //       <MemoryRouter>
  //         <Login />
  //       </MemoryRouter>
  //     </Provider>
  //   );

  //   const emailInput = screen.getByLabelText(/email/i);
  //   const submitButton = screen.getByRole('button', { type: /submit/i });

  //   fireEvent.change(emailInput, { target: { value: 'invalid-email' } });
  //   fireEvent.click(submitButton);

  //   expect(screen.getByText(/Invalid email or password/i)).toBeInTheDocument();
  // });
});
