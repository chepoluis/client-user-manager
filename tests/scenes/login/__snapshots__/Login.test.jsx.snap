// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests in <Login /> should match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-194fdgp"
  >
    <div
      class="MuiBox-root css-b5mc0i"
    >
      <div
        class="MuiBox-root css-0"
      >
        <h2
          class="MuiTypography-root MuiTypography-h2 css-be5vtk-MuiTypography-root"
        >
          LOGIN
        </h2>
        <h5
          class="MuiTypography-root MuiTypography-h5 css-b1sixc-MuiTypography-root"
        >
          Enter credentials to login
        </h5>
      </div>
      <div
        class="MuiBox-root css-0"
      />
    </div>
    <form>
      <div
        class="MuiBox-root css-ba4bgi"
      >
        <div
          class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-1vd3lbb-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-filled MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-filled css-o943dk-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="true"
            for=":r0:"
            id=":r0:-label"
          >
            Email
          </label>
          <div
            class="MuiInputBase-root MuiFilledInput-root MuiFilledInput-underline MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-2y464i-MuiInputBase-root-MuiFilledInput-root"
            data-testid="email-helper-text"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiFilledInput-input css-10botns-MuiInputBase-input-MuiFilledInput-input"
              id=":r0:"
              name="email"
              type="text"
              value="jonsnow@gmail.com"
            />
          </div>
        </div>
        <div
          class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-1vd3lbb-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-filled MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-filled css-o943dk-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="true"
            for=":r1:"
            id=":r1:-label"
          >
            Password
          </label>
          <div
            class="MuiInputBase-root MuiFilledInput-root MuiFilledInput-underline MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-2y464i-MuiInputBase-root-MuiFilledInput-root"
            data-testid="password-helper-text"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiFilledInput-input css-10botns-MuiInputBase-input-MuiFilledInput-input"
              id=":r1:"
              name="password"
              type="password"
              value="12333434"
            />
          </div>
        </div>
      </div>
      <div
        class="MuiBox-root css-1xaekgw"
      >
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1uydgkw-MuiGrid-root"
          style="margin-bottom: 10px;"
        >
          <div
            class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation0 MuiAlert-root MuiAlert-standardError MuiAlert-standard css-1hdyuqk-MuiPaper-root-MuiAlert-root"
            role="alert"
          >
            <div
              class="MuiAlert-icon css-1ytlwq5-MuiAlert-icon"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
                data-testid="ErrorOutlineIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M11 15h2v2h-2zm0-8h2v6h-2zm.99-5C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2zM12 20c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
                />
              </svg>
            </div>
            <div
              class="MuiAlert-message css-1pxa9xg-MuiAlert-message"
            />
          </div>
        </div>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSecondary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedSecondary MuiButton-sizeMedium MuiButton-containedSizeMedium css-ech2n8-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="submit"
        >
          Login
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </form>
  </div>
</DocumentFragment>
`;
